Theme Showcase: Visiting the showcase while logged in forces you to select a site.

mreishus opened the pull request:
#### Changes proposed in this Pull Request

* Visiting /themes while logged in forces you to select a site

#### Testing instructions

* Visit `/themes` while logged out
  *  Should still see the themes showcase
* Visit `/themes` while logged out and javascript is off
  *  Should still see the themes showcase (Server side rendered)
* Visit `/themes` while logged in
  * You should be forced to choose a site to proceed, and end up on a url like `/themes/<siteurl>`.

#### Dev Issues

Note: I am seeing some weirdness on the site selection page. However, I believe it is an artifact of isometric rendered pages (server side + client side) in the current dev environment.  Screenshot:

![2021-05-07_16-03](https://user-images.githubusercontent.com/937354/117508241-df61c000-af4d-11eb-8292-409badc08c89.png)

I don't believe it is related to this branch, because I see the same thing on the current trunk commit with no changes:

![2021-05-07_13-58](https://user-images.githubusercontent.com/937354/117508272-ebe61880-af4d-11eb-95b5-24c46317a5f1.png)

Notice production renders fine, but in dev, the same page under the same conditions does not render properly. I believe there is some sort of bug when the client side render 'takes over' the server side render.  

In short, without any changes in this PR, I am seeing `/themes` render quite differently in prod vs. localhost dev while logged in. Anyone know any more about this? (`server-side-rendering.md` mentioned these gurus: @ehg @mcsf @ockham @seear)

Edit: Reasoning for this discrepancy is explained here: pd2qbF-P-p2

**To temporarily fix the issue for localhost testing, also apply this patch to your local installation**:
```diff
diff --git a/client/controller/web-util.js b/client/controller/web-util.js
index 55d62fe50e..20940e794b 100644
--- a/client/controller/web-util.js
+++ b/client/controller/web-util.js
@@ -8,5 +8,5 @@ export function render( context ) {
 }
 
 export function hydrate( context ) {
-       ReactDom.hydrate( context.layout, document.getElementById( 'wpcom' ) );
+       ReactDom.render( context.layout, document.getElementById( 'wpcom' ) );
 }
```
A larger fix for this is in the works in #52731

Related to #14896

mreishus commented:
This PR is causing the e2e tests to fail, but in a unique way: They don't even build. This seems to be repeatable and unique to this branch, that is, it represents a real problem with this changeset. Currently unsure why.
```
  #15 18.87 [fallback ] <s> [webpack.Progress] 37% building 0/4 entries 16100/16162 dependencies 2520/4130 modules
09:43:08
  #15 18.95 Error: package not found: react-native, dependency of ./node_modules/react-native-url-polyfill
09:43:08
  #15 18.95     at processPackage (/calypso/bin/copy-production-modules.js:99:10)
09:43:08
  #15 18.95     at processPackage (/calypso/bin/copy-production-modules.js:109:3)
09:43:08
  #15 18.95     at processPackages (/calypso/bin/copy-production-modules.js:38:3)
09:43:08
  #15 18.95     at Object.<anonymous> (/calypso/bin/copy-production-modules.js:21:26)
09:43:08
  #15 18.95     at Module._compile (internal/modules/cjs/loader.js:1063:30)
09:43:08
  #15 18.95     at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
09:43:08
  #15 18.95     at Module.load (internal/modules/cjs/loader.js:928:32)
09:43:08
  #15 18.95     at Function.Module._load (internal/modules/cjs/loader.js:769:14)
09:43:08
  #15 18.95     at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:72:12)
09:43:08
  #15 18.95     at internal/main/run_main_module.js:17:47
09:43:08
  #15 18.96 error Command failed with exit code 1.
```

mreishus commented:
I'm able to recreate it locally by running `docker build --build-arg workers=8 --build-arg node_memory=32768 --build-arg use_cache=true -f Dockerfile .`.  However, I did need to set this up in my environment, to access the special a8c proxy which is used to contact our registry.

Create directory+file `/etc/systemd/system/docker.service.d/http-proxy.conf`:
```
[Service]
Environment="HTTP_PROXY=socks5://127.0.0.1:8080/"
Environment="HTTPS_PROXY=socks5://127.0.0.1:8080/"
```
Then `systemctl daemon-reload && systemctl restart docker` as root.

mreishus commented:
The docker build starts breaking once I add this line:
```
		return sites( context, next );
```
to
```
export function loggedIn( context, next ) {
```

Not sure why.

mreishus commented:
Cause (somewhat) discovered: Importing the sites middleware from `client/my-sites/controller.js` causes that file to be processed by `bin/copy-production-modules.js` during the docker build process. `client/my-sites/controller.js` includes `import { removeQueryArgs } from '@wordpress/url';`. 

`@wordpress/url` depends on `react-native-url-polyfill`, which depends on `react-native`, which cannot be included during this portion of the build process and causes an error. So I need a way to bring in the sites middleware without using any files that import `@wordpress/url` (or any other files that might have a dependency chain on react-native).

mreishus commented:
Fixed here:
`[update/themes-forces-site-selection 04e26f0a0d] Extract sites() middleware to its own file` 04e26f0a0d
Ready for re-review.

kwight commented:
Great, works as described. I'm not the best to review all the JS, but your explanation of extracting the middleware makes sense.

I saw one console error which isn't visible on prod (the first "Visit /themes while logged out" case), but prod looked so laughably different I'm not sure it's a concern. 🙃 

<img width="999" alt="Screen Shot 2021-05-11 at 2 09 00 PM" src="https://user-images.githubusercontent.com/349751/117886226-ba2dc400-b263-11eb-866f-fbedb7c087d2.png">

There's one seemingly related test still failing; I wonder if it needs a site to be specified (`[WPCOM] Themes: All sites (mobile): Preview a theme`).

mreishus commented:
>I saw one console error which isn't visible on prod (the first "Visit /themes while logged out" case)

Seeing the same error message. The mechanism makes sense:

*  `aria-controls` inside search [is rendered with an instanceId](https://github.com/Automattic/wp-calypso/blob/c751aff08fd72e7e9f7018e1fc7ceb615f1e9ee2/client/components/search/index.jsx#L349), which [comes from a random uuid](https://github.com/Automattic/wp-calypso/blob/c751aff08fd72e7e9f7018e1fc7ceb615f1e9ee2/client/components/search/index.jsx#L103). 
* The server side rendered version will create one random uuid, and the client hydrated version will create a second random uuid, which don't match, but [hydrate() expects the rendered content to be identical between the server and client](https://reactjs.org/docs/react-dom.html#hydrate).

 I'm not sure how to fix, or why the error isn't happening when I visit https://wordpress.com/themes while logged out; this does the same SSR Html -> Client Hydrate progression. Checking on production, I looked at the initial server rendered html sent over the wire with the network tab and saw `<div role="button" class="search__icon-navigation" aria-controls="search-component-43f07e51-fc52-4b53-84d3-955e8640762e"`, then checked `$(".search__icon-navigation")` in the console and saw the exact same uuid in aria-controls! I don't know why this matches; I would have expected the client side hydration to take over and choose a different UUID. Maybe @jsnajdr can advise if this is something we need to worry about or how to fix?

>There's one seemingly related test still failing; I wonder if it needs a site to be specified ([WPCOM] Themes: All sites (mobile): Preview a theme).

This makes sense, I'll check out the tests and look to update.

mreishus committed:
Update e2e tests after multiSite themes was removed

mreishus commented:
**Updated tests**

- `wp-theme-multisite-spec.js` used to have two tests: Previewing and activating. Both used an account with multiple sites.
  - Previewing works completely the same except it starts out by visiting `/themes/mysite.word press.com` instead of `/themes`.
  - Activating used to visit `/themes`, press activate, then look for a site selector and choose a site there. Now it visits `/themes/mysite.word press.com`, presses activate, and looks for a success modal.
 
* I wanted to add a third test that visits `/themes/ and checks that the "Choose a site" page loads, but I had trouble getting the e2e tests to work against localhost. Asked in slack for help.  This code is currently commented out.

* Working on the tests made me realize I should probably test going to `/themes` with an account that only has one site. I'll check that out tomorrow.

* The test file is still called `multisite-spec.js` .. unsure if this needs changing.

jsnajdr commented:
If you're removing the multi-site Themes view completely, you can also remove the `ThemesSiteSelectorModal` component and likely some others, too.

jsnajdr commented:
> aria-controls inside search is rendered with an instanceId, which comes from a random uuid.

The random uuids were introduced by @tyxla in #52673 (they were incrementing counters from lodash before) and since then, the IDs are guaranteed to be different in the client and server renders.

Before that, both IDs were usually `0` and they were the same only by accident. Server called `_.uniqueId` only at that one place, got `0`, and then cached the response and returned the cached markup with `0` on every `/themes` request. Client called `_.uniqueId` also only once (or at least the call in the Search component was the first one) and got `0`.

The best solution would be a context provider for instance IDs that keeps the ID counter local for a React tree. Then, if server and client really render the same thing, the IDs would be allocated in the same order and would be the same on both server and client: https://github.com/facebook/react/issues/5867#issuecomment-354774705

Good improvement idea for `useInstanceId` from `@wordpress/compose`.

jsnajdr commented on client/my-sites/themes/controller.jsx:
```diff
@@ -70,10 +70,13 @@ export function loggedIn( context, next ) {
+	} else {
+		// Force logged in user w/o a site to choose a site
+		return sites( context, next );
+	}
```
The trouble with importing `sites` could be solved by moving the `loggedIn` handler to a separate module that is never imported on server. I tried to do that in #52813.

Rendering the site selector can be done, arguably more elegantly, with two distinct route handlers:
```js
page( '/themes', siteSelection, sites, render );
page( '/themes/:site', siteSelection, loggedIn, navigation, render );
```

This approach keeps each handler focused on exactly one task, and prevents rendering sidebar (`navigation`) in the site selection UI, which is currently a bug in this PR:

<img width="832" alt="Screenshot 2021-05-13 at 11 39 45" src="https://user-images.githubusercontent.com/664258/118108272-1584c100-b3e0-11eb-8cbd-25a4d536d45a.png">

mreishus commented:
Next week: Needs rebasing off of #52813 and re-examination

mreishus committed:
Visiting /themes while logged in forces you to select a site - Alternate version

mreishus commented:
Reworked to use additional routes, one batch calling `loggedIn()`, another batch calling `sites()`. 
(The old method had  `loggedIn()` calling `sites()` sometimes, should be available on https://github.com/Automattic/wp-calypso/commit/ed3439843e121ccae220f316344ced7511ff64c6 and parents, now removed from this branch via rebasing. Worst case scenario I have a diff w/ the old changes on my HD).

Ready again for review: @jsnajdr @Automattic/serenity 

* I also need to check this with a user that only has one site.

jsnajdr commented on client/my-sites/themes/controller-logged-in.jsx:
```diff
@@ -28,8 +27,11 @@ export function loggedIn( context, next ) {
+	// Assuming context.params.site_id always exists
+	if ( context.params.site_id ) {
+		console.error( 'theme/loggedIn: Missing required parameter site_id' ); // eslint-disable-line no-console
+	}
```
You probably wanted to check `! context.params.site_id`, but I think this check is redundant anyway: the presence of `site_id` is guaranteed by the route definition, and can be removed as a debug message that's no longer needed.

mreishus replied:
Yea, it was supposed to be negative

>the presence of site_id is guaranteed by the route definition,

This is true. I was mostly thinking of the next developer to come around and work on routes. They might not realize that the `loggedIn()` function assumes that a `site_id` is present, start rearranging the routes and middleware, and violate an assumption that isn't obvious.

jsnajdr commented on client/my-sites/themes/index.web.js:
```diff
@@ -54,9 +61,19 @@ export default function ( router ) {
+			validateFilters,
+			siteSelection,
+			sites,
+			makeLayout
```
The `fetch` and `validate` handlers can be removed. We can show the site selector even for un-validated `vertical` and `filter` parameters, and verify them only after navigating to the `site_id`-enabled route.

jsnajdr replied:
We could also remove the `isLoggedIn` check from the code that installs the route handlers and unify them into a single handler set:
```js
router(
  routesWithSites,
  redirectToLoginIfLoggedOut,
  fetchAndValidateVerticalsAndFilters,
  siteSelection,
  loggedIn,
  navigation,
  makeLayout
);

router(
  routesWithoutSites,
  selectSiteIfLoggedIn,
  fetchAndValidateVerticalsAndFilters,
  loggedOut,
  makeLayout
);
```
- routes with sites are for logged-in users. Logged-out users are redirected to login.
- routes without sites are for logged-out users. Logged-in users are shown asite selector and then to a site-specific route.

Each handler has an "escape hatch" at the beginning and then proceeds to render the main case.

jsnajdr commented on client/my-sites/themes/index.web.js:
```diff
@@ -54,9 +61,19 @@ export default function ( router ) {
 	} else {
 		router(
-			routes,
+			routesWithSites.concat( routesWithoutSites ),
```
`routesWithSites` could have a dedicated handler here, which does a simple:
```js
router( routesWithSites, redirectToLoginRightAway );
```
If I'm logged out and go to a route with `site_id`, a redirect to login is guaranteed.

mreishus committed:
Rework to use only two route calls

mreishus committed:
Move console.error to documentation note

mreishus committed:
Remove alternate compose method comments

mreishus committed:
Remove comment

mreishus committed:
Remove themes-site-selector-modal

mreishus committed:
Remove site-selector-modal

mreishus committed:
Remove the SiteSelector used in e2e tests

mreishus committed:
Move selectSiteIfLoggedIn to controller-logged-in.jsx to fix docker builds

mreishus commented:
* Reworked the routes to work with only two calls to `router`
  * This ended up pushing the complexity inside the middleware calls, for example: 
```
	router(
		routesWithoutSites,
		selectSiteIfLoggedIn, //  <-- Only takes action if logged in
		fetchAndValidateVerticalsAndFiltersIfLoggedOut,  //  <-- Only takes action if logged out
		loggedOut,                                       //  <-- Only takes action if logged out
		makeLayout
	);
```
  * Alternate solution from the beginning of this PR here: https://github.com/Automattic/wp-calypso/pull/52984/files
  * I wasn't sure how to compose middleware ([link](https://github.com/Automattic/wp-calypso/blob/ce621f4998df927cdec8e88a8833d57335411118/client/my-sites/themes/validate-filters.js#L95-L113)) - I couldn't find the source code to `router()` so I don't know exactly what it's doing
  * Removed `<ThemesSiteSelectorModal`
  * Removed `<SiteSelectorModal` - was only used by  `<ThemesSiteSelectorModal`
  * Removed `SiteSelectorComponent` in e2e tests

mreishus commented:
E2e tests are showing 0 passed tests. Sounds like a build failure. However, I'm not able to interpret the error message I'm seeing in on teamcity:
```
  [INFO] [Magellan]   -> Magellan doesn't require a resource manager to schedule test run
16:04:37 [INFO] [Magellan] Searching for tests...
16:04:40 [ERROR] [Mocha Plugin] [
16:04:40  'Could not capture mocha tests. To debug, run the following command:\n' +
16:04:40    'MOCHA_CAPTURE_PATH=%s %s %s',
16:04:40  '/home/teamcity-1/buildAgent/work/c4a9d5b38c1dacde/test/e2e/temp/get_mocha_tests.json',
16:04:40  '/home/teamcity-1/buildAgent/work/c4a9d5b38c1dacde/node_modules/mocha/bin/mocha',
16:04:40  '--reporter-options tags=parallel --reporter /home/teamcity-1/buildAgent/work/c4a9d5b38c1dacde/packages/magellan-mocha-plugin/lib/suite_capture.js specs/specs-calypso'
16:04:40 ]
16:04:40 error Command failed with exit code 1.
```
Edit: Running an individual spec for me locally is working.

jsnajdr commented:
> I wasn't sure how to compose middleware

Method nr. 2 is the right one -- pass a nested `next` callback. `next` can be called asynchronously, or not at all --  the handler may decide it fully handler the route and skips the rest of the chain.

It should be possible to create a fancy `composeHandlers( [ h1, h2, h3 ] )` helper using `reduceRight` or similar. But I haven't figured it out in the 5 min timebox I had for the task 🙂

jsnajdr commented on client/my-sites/theme/main.jsx:
```diff
@@ -769,17 +768,7 @@ class ThemeSheet extends React.Component {
-		</ThemesSiteSelectorModal>
-	);
-} );
+const ConnectedThemeSheet = connectOptions( ( props ) => <ThemeSheet { ...props } /> );
```
Could be a more direct `connectOptions( ThemeSheet )`.

jsnajdr commented on client/my-sites/themes/controller-logged-in.jsx:
```diff
@@ -47,3 +48,13 @@ export function upload( context, next ) {
+	const currentUser = getCurrentUser( state );
+
+	if ( currentUser ) {
+		return sites( context, next );
```
You want to also call `siteSelection` handler before `sites`. It has no actual site selection to perform, but if the user has only single site, it will redirect from `/themes` to `/themes/that.site` right away, without showing the selection UI.

The handler also shouldn't call `next` in case it decides to render the site selector. It wants to render the site selector and terminate the handler chain.
```js
if ( ! isUserLoggedIn( state ) {
  next();
  return;
}

siteSelection( context, () => {
  sites( context, () => {
    makeLayout( context, () => {
      clientRender( context )
    } );
  } );
} );
```

jsnajdr commented on client/my-sites/themes/controller.jsx:
```diff
@@ -43,6 +44,13 @@ export function loggedOut( context, next ) {
 
+	// If logged in, do nothing:
+	const state = context.store.getState();
+	const currentUser = getCurrentUser( state );
```
We also have a specialized `isUserLoggedIn( state )` selector that returns a boolean.

jsnajdr replied:
This extra condition is not needed if the `selectSiteIfLoggedIn` handler terminates early and doesn't call `next`.

One way to think about it is that earlier handlers in the chain check pre-conditions for later handlers and proceed only if the pre-conditions are satisfied.

jsnajdr commented on client/my-sites/themes/validate-filters.js:
```diff
@@ -81,3 +83,33 @@ export function sortFilterTerms( context, terms ) {
+
+	if ( ! currentUser ) {
+		return next();
+	}
```
You don't need these logged-in/out checks if the `selectSiteIfLoggedIn` and `redirectLoggedOut` handlers terminate early and don't call `next`. `redirectLoggedOut` already behaves that way. `selectSiteIfLoggedIn` needs to be fixed.

Then both `fetchAndValidate` handlers can be merged into one.

jsnajdr commented on client/my-sites/themes/validate-filters.js:
```diff
@@ -81,3 +83,33 @@ export function sortFilterTerms( context, terms ) {
+
+	return fetchThemeFilters( context, () => {
+		return validateVertical( context, () => {
+			return validateFilters( context, next );
```
A little nitpick: the handlers don't return anything -- they pass control solely using the callbacks. You can just call `validateVertical` etc. without the `return` statement.

jsnajdr reviewed (commented):
This is really close to being shippable. I'm suggesting a few fixes.

mreishus committed:
Rework route handler escape hatches

mreishus committed:
ConnectedThemeSheet small refactor

mreishus commented:
Ah, I get it - handlers are under no obligation to call `next()`, so they can set up their own "escape hatches" of other handler chains to call.
Pushed an update, waiting to see if the docker build and e2e tests work.

mreishus commented:
Still getting 0 passed e2e tests and a build failure. 
```
10:02:52  [INFO] [Magellan] Creating temporary files at: /home/teamcity-3/buildAgent/work/c4a9d5b38c1dacde/test/e2e/temp
10:02:53  [INFO] [Magellan] Version:  11.0.10
10:02:53  [INFO] [Magellan] Use --help to list out all command options
10:02:53  [INFO] [Magellan] Loaded test framework from magellan.json:
10:02:53  [INFO] [Magellan]   @automattic/testarmada-magellan-mocha-plugin
10:02:53  [INFO] [Magellan] Loaded test executors from magellan.json:
10:02:53  [INFO] [Magellan]   testarmada-magellan-local-executor
10:02:53  [INFO] [Magellan] Enabled executors:
10:02:53  [INFO] [Magellan]   testarmada-magellan-local-executor
10:02:53  [INFO] [Magellan] Enabled bail strategy:
10:02:53  [INFO] [Magellan]   testarmada-magellan-never-bail-strategy:
10:02:53  [INFO] [Magellan]   -> Magellan never bails, all tests will be executed at least once
10:02:53  [INFO] [Magellan] Enabled resource strategy:
10:02:53  [INFO] [Magellan]   testarmada-magellan-no-resource-strategy:
10:02:53  [INFO] [Magellan]   -> Magellan doesn't require a resource manager to schedule test run
10:02:53  [INFO] [Magellan] Searching for tests...
10:02:56  [ERROR] [Mocha Plugin] [
10:02:56    'Could not capture mocha tests. To debug, run the following command:\n' +
10:02:56      'MOCHA_CAPTURE_PATH=%s %s %s',
10:02:56    '/home/teamcity-3/buildAgent/work/c4a9d5b38c1dacde/test/e2e/temp/get_mocha_tests.json',
10:02:56    '/home/teamcity-3/buildAgent/work/c4a9d5b38c1dacde/node_modules/mocha/bin/mocha',
10:02:56    '--reporter-options tags=parallel --reporter /home/teamcity-3/buildAgent/work/c4a9d5b38c1dacde/packages/magellan-mocha-plugin/lib/suite_capture.js specs/specs-calypso'
10:02:56  ]
10:02:56  error Command failed with exit code 1.
```

Guess that's the last thing to figure out.

kwight commented:
@Automattic/quality-squad Any ideas why e2e tests won't build for this PR?..

mreishus commented:
@Automattic/quality-squad @kwight I did eventually figure it out, it was due to how I was checking an element exists on the page:
```
import { By } from 'selenium-webdriver';
import * as driverHelper from '../lib/driver-helper';
				it( 'can click activate and see the success modal', async function () {
					await this.themesPage.clickPopoverItem( 'Activate' );
					const thanksModalShown = await driverHelper.isElementEventuallyLocatedAndVisible(
						driver,
						By.css( '.themes__thanks-modal' )
					);
					return assert(
						thanksModalShown,
						"The 'Thanks for Choosing Twenty Sixteen' modal was not displayed after activating"
					);
				} );
```
^ This causes the e2e tests to not run on teamcity (but it works fine for me locally) 
I assume it has to do with me importing from selenium directly.. any alternative method suggestions are appreciated!

mreishus commented:
This is the test error I see, I think it's not related to this branch
![2021-05-19_17-21](https://user-images.githubusercontent.com/937354/118892528-a2d78200-b8c6-11eb-8c53-06baf060fee8.png)

jsnajdr commented:
> handlers are under no obligation to call `next()`

Yes that's true. I was finally able to wire up a `composeHandler` helper:
```js
function composeHandlers( ...handlers ) {
  return ( context, next ) => {
    const it = handlers.values();
    function handleNext() {
      const nextHandler = it.next().value;
      if ( ! nextHandler ) {
        next();
      } else {
        nextHandler( context, handleNext );
      }
    }
    handleNext();
  };
}
```
(instead of `reduceRight` or `reduce` I ended up using recursion)

It could be used in the `selectSiteIfLoggedIn` handler like this:
```js
composeHandlers( siteSelection, sites, makeLayout, clientRender )( context );
```

Or for another handler:
```js
export const fetchAndValidateVerticalsAndFilters = composeHandlers(
  fetchThemeFilters,
  validateVertical,
  validateFilters
);
```

jsnajdr reviewed (approved):
Looks good and works well :+1: I think this ended up being a very elegant refactoring that clarifies the very complex route handling in `/themes`.

kwight commented:
Great work @mreishus ! 🚀

mreishus commented:
Rebasing to solve conflicts with some e2e test refactors and renaming on trunk, then looking how the e2e test run now. I can't help but feel that a merge based workflow inside this PR would have made this much easier (one conflict resolution vs one at each step), and that more history would be preserved.

mreishus commented:
I ended up with 
```
import { By } from 'selenium-webdriver';
import * as driverHelper from '../lib/driver-helper';
``` 

which I definitely don't want - the rebase did not apply my changes to the renamed files (IE, I would have expected https://github.com/Automattic/wp-calypso/commit/7349b787320b10df4cbe3d1652054f0e15dffa60 to be applied to the renamed `wp-theme__multisite-activate-spec.js` - but possibly not, because my previous change to `wp-theme__multisite-activate-spec.js` was implemented as a rebase conflict resolution) . Investigating.

Edit: Fixed.

mreishus committed:
Comment out test crashing CI e2e tests

mreishus committed:
Tabs

mreishus commented:
Is Pull Request / ICFY / Repo Gardening checks are stuck due to a github issue.
